core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -544,6 +544,10 @@ final Exception checkIfDocumentFailureOrThrow(final Operation operation, final E
             // and set the error in operation.setFailure. In case of environment related errors, the failure
             // is bubbled up
             isDocumentFailure = maybeFailEngine(operation.operationType().getLowercase(), failure) == false;
+            if (failure instanceof EngineClosedException) {
```

Comment:
I am not sure what makes a `EngineClosedException` not a `isDocumetnFailure` but `AlreadyClosedException` does. I think we use the boolean returned from `maybeFailEngine` in the wrong way here. It almost seems like that we need some state handling on `Operation` that tells us if the document actually failed to index or if we didn't try or if we tried, succeeded but after success we failed due to concurrent closing. This entire exception based decision making here seems error prone to me.